mayastor/src/persistent_store.rs

```diff
@@ -0,0 +1,250 @@
+            false => format!("{}:{}", endpoint, DEFAULT_PORT),
+        }
+    }
+
+    /// Connect to etcd as the backing store.
+    /// A connection to the store will be attempted continuously until
+    /// successful. This is necessary as the backing store is essential to the
+    /// operation of Mayastor across restarts.
+    async fn connect_to_backing_store(endpoint: &str) -> Etcd {
+        let mut output_err = true;
+        loop {
+            match Etcd::new(endpoint).await {
+                Ok(store) => {
+                    info!("Connected to etcd on endpoint {}", endpoint);
+                    return store;
+                }
+                Err(_) => {
+                    if output_err {
+                        // Only output the error on first failure to prevent
+                        // flooding the logs.
+                        error!(
+                            "Failed to connect to etcd on endpoint {}",
+                            endpoint
+                        );
+                        output_err = false;
+                    }
+                    tokio::time::sleep(Duration::from_secs(1)).await;
+                }
+            }
+        }
+    }
+
+    /// Put a key-value in the store.
+    pub async fn put(
+        key: &impl StoreKey,
+        value: &impl StoreValue,
+    ) -> Result<(), StoreError> {
+        let put_value = serde_json::to_value(value)
+            .expect("Failed to convert value to a serde_json value");
+        let key_string = key.to_string();
+        let value_clone = put_value.clone();
+
+        let rx = Self::execute_store_op(async move {
+            info!(
+                "Putting key {}, value {} in store.",
+                key_string,
+                value_clone.to_string()
+            );
+            match Self::backing_store()
+                .put_kv(&key_string, &value_clone)
+                .await
+            {
+                Ok(_) => {
+                    info!(
+                        "Successfully put key {}, value {} in store.",
+                        key_string,
+                        value_clone.to_string()
+                    );
+                    Ok(None)
+                }
+                Err(e) => Err(e),
+            }
+        });
+
+        let result = rx.await.context(PutWait {
+            key: key.to_string(),
+            value: put_value.to_string(),
+        })?;
+        result.map(|_| ())
+    }
+
+    /// Retrieve a value, with the given key, from the store.
+    pub async fn get(key: &impl StoreKey) -> Result<Value, StoreError> {
+        let key_string = key.to_string();
+        let rx = Self::execute_store_op(async move {
+            info!("Getting key {} from store.", key_string);
+            match Self::backing_store().get_kv(&key_string).await {
+                Ok(value) => {
+                    info!("Successfully got key {}", key_string);
+                    Ok(Some(value))
```

Comment:
can't this be just Ok(value) and then below on 157 you don't need to unwrap?

Reply:
As discussed, changed the execute store op function to be generic in its return type.